sqlmesh/core/engine_adapter/spark.py

```diff
@@ -50,37 +51,137 @@ def _use_spark_session(self) -> bool:
+            exp.DataType.Type.MAP: spark_types.MapType,
+            exp.DataType.Type.STRUCT: spark_types.StructType,
+        }
+
+        SPARK_PRIMITIVE_MAPPING = {v: k for k, v in PRIMITIVE_MAPPING.items()}
+        SPARK_COMPLEX_TYPE_MAPPING = {v: k for k, v in COMPLEX_TYPE_MAPPING.items()}
+
+        def sqlglot_complex_to_spark_complex(complex_type: exp.DataType) -> spark_types.DataType:
+            is_struct = complex_type.is_type(exp.DataType.Type.STRUCT)
+            expressions = []
+            for column_def in complex_type.expressions:
+                col_name = column_def.this.name if is_struct else None
+                data_type = column_def.args["kind"] if is_struct else column_def
+                primitive_func = PRIMITIVE_MAPPING.get(data_type.this)
+                type_func = (
+                    primitive_func
+                    if primitive_func
+                    else partial(sqlglot_complex_to_spark_complex, data_type)
+                )
+                if is_struct:
+                    expressions.append(spark_types.StructField(col_name, type_func()))
+                else:
+                    expressions.append(type_func())
+            klass = COMPLEX_TYPE_MAPPING[complex_type.this]
+            return klass(*([expressions] if is_struct else expressions))
+
+        def spark_complex_to_sqlglot_complex(
+            complex_type: t.Union[
+                spark_types.StructType, spark_types.ArrayType, spark_types.MapType
+            ],
+        ) -> exp.DataType:
+            def get_fields(
+                complex_type: t.Union[
+                    spark_types.StructType, spark_types.ArrayType, spark_types.MapType
                 ]
+            ) -> t.Sequence[spark_types.DataType]:
+                if isinstance(complex_type, spark_types.StructType):
+                    return complex_type.fields
+                if isinstance(complex_type, spark_types.ArrayType):
+                    return [complex_type.elementType]
+                if isinstance(complex_type, spark_types.MapType):
+                    return [complex_type.keyType, complex_type.valueType]
+                raise SQLMeshError(f"Unsupported complex type: {complex_type}")
+
+            def get_expression(
+                col_name: str, dtype: t.Union[exp.DataType, exp.DataType.Type]
+            ) -> t.Union[exp.ColumnDef, exp.DataType]:
+                kind = dtype if isinstance(dtype, exp.DataType) else exp.DataType(this=dtype)
+                if col_name:
+                    return exp.ColumnDef(this=exp.to_identifier(col_name), kind=kind)
+                return kind
+
+            expressions = []
+            fields = get_fields(complex_type)
+            for field in fields:
+                col_name = field.name if hasattr(field, "name") else None
+                spark_data_type = field.dataType if hasattr(field, "dataType") else field
+                primitive = SPARK_PRIMITIVE_MAPPING.get(type(spark_data_type))
```

Comment:
i feel like this code could be cleaner, just need two methods, 

sqlglot_to_spark
spark_to_sqlglot

def sqlglot_to_spark:
   if isinstance(struct):
       ..
   if primitive type:
      ..

Reply:
Yeah that is another way to do it. I'm indifferent either way so I can do that.